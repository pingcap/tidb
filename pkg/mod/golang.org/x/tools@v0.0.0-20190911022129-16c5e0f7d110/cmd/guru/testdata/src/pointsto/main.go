package main

// Tests of 'pointsto' query.
// See go.tools/guru/guru_test.go for explanation.
// See pointsto.golden for expected query results.

const pi = 3.141 // @pointsto const "pi"

var global = new(string) // NB: ssa.Global is indirect, i.e. **string

func main() {
	livecode()

	// func objects
	_ = main   // @pointsto func-ref-main "main"
	_ = (*C).f // @pointsto func-ref-*C.f "..C..f"
	_ = D.f    // @pointsto func-ref-D.f "D.f"
	_ = I.f    // @pointsto func-ref-I.f "I.f"
	var d D
	var i I
	_ = d.f // @pointsto func-ref-d.f "d.f"
	_ = i.f // @pointsto func-ref-i.f "i.f"

	// var objects
	anon := func() {
		_ = d.f // @pointsto ref-lexical-d.f "d.f"
	}
	_ = anon   // @pointsto ref-anon "anon"
	_ = global // @pointsto ref-global "global"

	// SSA affords some local flow sensitivity.
	var a, b int
	var x = &a // @pointsto var-def-x-1 "x"
	_ = x      // @pointsto var-ref-x-1 "x"
	x = &b     // @pointsto var-def-x-2 "x"
	_ = x      // @pointsto var-ref-x-2 "x"

	i = new(C) // @pointsto var-ref-i-C "i"
	if i != nil {
		i = D{} // @pointsto var-ref-i-D "i"
	}
	print(i) // @pointsto var-ref-i "\\bi\\b"

	m := map[string]*int{"a": &a}
	mapval, _ := m["a"] // @pointsto map-lookup,ok "m..a.."
	_ = mapval          // @pointsto mapval "mapval"
	_ = m               // @pointsto m "m"

	if false {
		panic(3) // @pointsto builtin-panic "panic"
	}

	// NB: s.f is addressable per (*ssa.Program).VarValue,
	// but our query concerns the object, not its address.
	s := struct{ f interface{} }{f: make(chan bool)}
	print(s.f) // @pointsto var-ref-s-f "s.f"
}

func livecode() {} // @pointsto func-live "livecode"

func deadcode() { // @pointsto func-dead "deadcode"
	// Pointer analysis can't run on dead code.
	var b = new(int) // @pointsto b "b"
	_ = b
}

type I interface {
	f()
}

type C int
type D struct{}

func (c *C) f() {}
func (d D) f()  {}
