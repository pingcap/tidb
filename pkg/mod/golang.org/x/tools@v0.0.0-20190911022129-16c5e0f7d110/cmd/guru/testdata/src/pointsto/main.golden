-------- @pointsto const --------

Error: pointer analysis wants an expression of reference type; got untyped float
-------- @pointsto func-ref-main --------
this func() may point to these objects:
	pointsto.main

-------- @pointsto func-ref-*C.f --------
this func() may point to these objects:
	(*pointsto.C).f

-------- @pointsto func-ref-D.f --------
this func() may point to these objects:
	(pointsto.D).f

-------- @pointsto func-ref-I.f --------

Error: func (pointsto.I).f() is an interface method
-------- @pointsto func-ref-d.f --------
this func() may point to these objects:
	(pointsto.D).f

-------- @pointsto func-ref-i.f --------

Error: func (pointsto.I).f() is an interface method
-------- @pointsto ref-lexical-d.f --------
this func() may point to these objects:
	(pointsto.D).f

-------- @pointsto ref-anon --------
this func() may point to these objects:
	pointsto.main$1

-------- @pointsto ref-global --------
this *string may point to these objects:
	new

-------- @pointsto var-def-x-1 --------
this *int may point to these objects:
	a

-------- @pointsto var-ref-x-1 --------
this *int may point to these objects:
	a

-------- @pointsto var-def-x-2 --------
this *int may point to these objects:
	b

-------- @pointsto var-ref-x-2 --------
this *int may point to these objects:
	b

-------- @pointsto var-ref-i-C --------
this I may contain these dynamic types:
	*C, may point to:
		new

-------- @pointsto var-ref-i-D --------
this I may contain these dynamic types:
	D

-------- @pointsto var-ref-i --------
this I may contain these dynamic types:
	*C, may point to:
		new
	D

-------- @pointsto map-lookup,ok --------

Error: pointer analysis wants an expression of reference type; got (*int, bool)
-------- @pointsto mapval --------
this *int may point to these objects:
	a

-------- @pointsto m --------
this map[string]*int may point to these objects:
	makemap

-------- @pointsto builtin-panic --------

Error: pointer analysis wants an expression of reference type; got ()
-------- @pointsto var-ref-s-f --------
this interface{} may contain these dynamic types:
	chan bool, may point to:
		makechan

-------- @pointsto func-live --------

Error: pointer analysis did not find expression (dead code?)
-------- @pointsto func-dead --------

Error: pointer analysis did not find expression (dead code?)
-------- @pointsto b --------

Error: pointer analysis did not find expression (dead code?)
